xml_converter/src/xml_converter.cpp

```diff
@@ -64,83 +65,85 @@ void write_xml_file(string xml_filepath, map<string, Category>* marker_categorie
         throw std::error_code();
     }
-
+    // Converts map<string, Category> to Waypoint::Categories
+    waypoint::Waypoint all_categories;
+    std::set<string> proto_categories;
     for (const auto& category : *marker_categories) {
         waypoint::Category proto_category = category.second.as_protobuf();
         all_categories.add_category()->CopyFrom(proto_category);
         create_proto_categories(proto_category, &proto_categories);
     }
-
+    // Converts vector<Parseable*> to Waypoint::Icon and Waypoint::Trail
+    waypoint::Waypoint proto_pois;
+    std::set<int> map_ids;
     for (const auto& parsed_poi : *parsed_pois) {
         if (parsed_poi->classname() == "POI") {
             Icon* icon = dynamic_cast<Icon*>(parsed_poi);
             waypoint::Icon poi = icon->as_protobuf();
             proto_pois.add_icon()->CopyFrom(poi);
             map_ids.insert(icon->map_id);
         }
-        if (parsed_poi->classname() == "Trail") {
+        else if (parsed_poi->classname() == "Trail") {
             Trail* trail = dynamic_cast<Trail*>(parsed_poi);
             waypoint::Trail poi = trail->as_protobuf();
             proto_pois.add_trail()->CopyFrom(poi);
             map_ids.insert(trail->map_id);
         }
     }
 
+    waypoint::Waypoint output_message;
+    std::set<string> category_includes;
     for (int map_id : map_ids) {
         ofstream outfile;
         string output_filepath = proto_filepath + "/" + to_string(map_id) + ".data";
         outfile.open(output_filepath, ios::out | ios::binary);
         output_message.CopyFrom(all_categories);
 
         for (int i = 0; i < proto_pois.icon_size(); i++) {
-            vector<string> split_categories = split(proto_pois.icon(i).category().name(), ".");
             if (proto_pois.icon(i).map_id() == map_id) {
-                auto category = proto_categories.find(split_categories[split_categories.size() - 1]);
+                vector<string> split_categories = split(proto_pois.icon(i).category().name(), ".");
+                auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
                 if (category == proto_categories.end()) {
                     cout << "Unknown MarkerCategory " << proto_pois.icon(i).category().name() << endl;
                 }
                 else {
-                    for (string split_category : split_categories) {
-                        category_includes.insert(split_category);
+                    // This loop adds the name of the category and all of it's parents
+                    string name;
+                    for (unsigned int j = 0; j < split_categories.size(); j++) {
+                        name += split_categories[j];
+                        category_includes.insert(name);
+                        name += ".";
                     }
                 }
                 output_message.add_icon()->CopyFrom(proto_pois.icon(i));
```

Comment:
we seem to call `add_icon()` on `proto_pois` only to then call `add_icon()` to copy the data out of `proto_pois`. Is there a reason we need `proto_pois`. This goes for the other `waypoint::Waypoint` variables that each hold a different element. This seems wasteful on the surface because all the data is being copied into a `waypoint::Waypoint` only to then be copied into another `waypoint::Waypoint` slightly later. While I believe the pruning argument for "copy and delete" holds water this "copy and then copy again" needs a robust case to be made as to why the data is not just copied once into the final location.